core/src/main/scala/chisel3/package.scala

```diff
@@ -210,26 +211,140 @@ package object chisel3 {
+     * /** The formatted string interpolator.
+     *  The `f` method works by assembling a format string from all the `parts` strings and using
+     *  `java.lang.String.format` to format all arguments with that format string. The format string is
+     *  obtained by concatenating all `parts` strings, and performing two transformations:
+     *   1. Let a _formatting position_ be a start of any `parts` string except the first one.
+     *      If a formatting position does not refer to a `%` character (which is assumed to
+     *      start a format specifier), then the string format specifier `%s` is inserted. This means any string with no format specifiers should behave same as p interpolator.
+     *   2. If format specifiers specifically target Chisel types - Like Data/Bits -  specific internal methods are called on that argument (like FirrtlFormat)
+     *   3. If format specifers target regular scala data types - java.lang.String.format method is called with specifier and argument to get the formatted string.
+     *
+     * Yet to document all the corener cases which this code tries to cover - but here are few of them
+     * 1. Allow calling %f or %2.2f (or something similar) on Int type data (Int, Short, Long) - this requires doing an explicit cast of the data before calling String.format
+     * 2. For bits - if no format specifier given (%s i.e) call .toPrintable to pick the default one. For regular data - let the String.format handle canonical case of %s too.
+     * Still testing this code - I am pretty sure this can be refactored and made more efficient / readable. WIP.
+     *
+     * /
+     */
+     */
+    def cf(args: Any*): Printable = {
+
+      // Handle literal % 
+      // Takes the part string - 
+      // - this is assumed to not have any format specifiers - already handled / removed before calling this function. 
+      // Only thing present is literal % if any which should ideally be with %%. 
+      // If not - then flag an error. 
+      // Return seq of Optional Printables (either PString or Percent or both - nothing else)
+      def PercentSplitter(s : String) : Seq[Option[Printable]] = {
+        if(s.isEmpty()) return Seq(Some(PString("")))
+        var iter = 0
+        var curr_start = 0
+        val buf = mutable.ListBuffer.empty[Option[Printable]]
+        while(iter < s.size) {
+          if(s(iter) == '%' ) {
+            if(iter >= s.size - 1 || s(iter + 1 ) != '%') {
+              throw new UnknownFormatConversionException("Un-escaped % found")
+            }
+            if(curr_start < iter) {
+              buf += Some(PString(s.substring(curr_start,iter)))
+            }
+            buf += Some(Percent)
+            curr_start = iter + 2
+            iter += 2
+          }
+          else {
+            iter += 1
+          }
+        }
+
+        // Handle the tail
+        if(curr_start < iter) {
+          buf += Some(PString(s.substring(curr_start,iter)))
+        }
+        buf.toSeq
```

Comment:
```suggestion
      def percentSplitter(s : String): Seq[Printable] = {
        val pieces = s.split("%%").toList.flatMap { p =>
          if (p.contains('%')) throw new UnknownFormatConversionException("Un-escaped % found")
          // Wrap in PString and intersperse the escaped percentages
          Seq(Percent, PString(p))
        }
        pieces.tail // Don't forget to drop the extra percent we put at the beginning
```
This can be a lot simpler unless I'm missing something. Also I'm not sure why you're using `Option[Printable]` when you only have have `Some`. Also, method names should start with lower case letters

Reply:
Thanks! Incorporated - had to make some minor changes for corner cases (like only if %% exists in whole string)